File: easy/trees/python/102-level-order-traversal.py
from typing import Optional, List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        q = []
        res = []
        if root:
            res.append([root.val])
            q.append(root)
        while q:
            children = []
            for node in q:
                left = node.left
                right = node.right

                if left:
                    children.append(left)
                if right:
                    children.append(right)
            if children:
                res.append([child.val for child in children])
                q = children
            else:
                break

        return res

File: easy/trees/cpp/102-level-order-traversal.cpp
#include "../../../stdc++.h"
using std::queue;
using std::vector;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    vector<vector<int>> levelOrder(TreeNode *root)
    {
        if (!root)
        {
            return {};
        }
        vector<vector<int>> ans;
        queue<TreeNode *> q;

        q.push(root);

        while (!q.empty())
        {
            vector<int> row;

            int size = q.size();
            while (size > 0)
            {
                TreeNode *node = q.front();
                q.pop();

                row.push_back(node->val);
                if (node->left != nullptr)
                {
                    q.push(node->left);
                }
                if (node->right != nullptr)
                {
                    q.push(node->right);
                }
                size--;
            }
            ans.push_back(row);
        }
        return ans;
    }
};
